### PR TITLE
MueLu: uncoupled aggregation refactor

### LINKED ISSUE #1366: MueLu uncoupled aggregation not threaded
Trilinos was built with kokkos node type OpenMP.
Looked at thread scaling for a single KNL.
Running with 1 MPI process and increasing number of threads from 1 to 64 (so have one thread per core).

Although the matrix-matrix multiplication (kokkoskernels SPGEMM) thread scales well,
the uncoupled aggregation does not thread scale at all.
For 1 thread, uncoupled aggregation accounts for 1/3 of the MueLu setup time,
and 40% of the uncoupled aggregation time (so 1/8 of MueLu setup time)
is due to the use of std::sort, which is not threaded (either lines 916 or 999 or both of muelu/src/Graph/MueLu_CoalesceDropFactory_def.hpp).
As increase the number of threads, and because R*A*P thread scales well, std::sort becomes a larger fraction of the time.
By the time get to 64 threads, matrix-matrix multiplication is only ~7% of MueLu setup time and uncoupled aggregation 90% of the time.

times in seconds

t  MLsetup MLrap  MLagg MLcoal MLagg1 MLagg2a MLagg3a 
1   259.4  174.9   78.2   58.2    8.0    3.8    6.5   
2   173.0   89.9   78.1   58.1    8.0    3.8    6.5  
4   130.1   47.4   78.2   58.2    8.0    3.8    6.5  
8   107.6   25.5   78.2   58.2    8.0    3.8    6.5  
16   96.4   14.4   78.2   58.2    8.0    3.8    6.5
32   90.8    9.0   78.2   58.2    8.0    3.8    6.5
64   88.0    6.2   78.2   58.2    8.0    3.8    6.5

legend:
MLsetup="MueLu Hierarchy: Setup (total)"
MLrap="MueLu RAPFactory: Computing Ac" (not "total")
MLagg="UncoupledAggregationFactory: Build (total)"
MLcoal="MueLu CoalesceDropFactory: Build" (not "total")
MLagg1="MueLu AggregationPhase1Algorithm: BuildAggregates"
MLagg2a="MueLu AggregationPhase2aAlgorithm: BuildAggregates"
MLagg2b="MueLu AggregationPhase2bAlgorithm: BuildAggregates"

What would be the schedule for threading uncoupled aggregation?

### PR DESCRIPTION
**Pull request by lucbv**
This pull request is providing a refactor of the uncoupled aggregation algorithms to leverage kokkos capabilities as well as the work done on distance 2 graph coloring.

@trilinos/muelu 

## Description
The interface of aggregation algorithms now requires the number of colors and color assignment from the distance 2 coloring graph algorithm as input.
At the moment the coloring algorithm is called in phase 1 of aggregation but this will be moved to the UncoupledAggregationFactory.
Another item to be modified is the aggStat vector that is created as an `std::vector<unsigned>` in UncoupledAggregationFactory 

## Motivation and Context
This provides an implementation of the uncoupled aggregation algorithms that is architecture portable and that should improve performance on manycore/GPU machines.

## Related Issues
* Closes #1366, #2430
* Blocks  #1219 

## How Has This Been Tested?
`muelu/test/scaling/Driver.cpp` is modified within this pull request and is used as the primary vehicle for tests but more tests needs to be added since the driver will return to its normal form upon merge of this pull request.

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ ] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase1Algorithm_kokkos_def.hpp [hunk 1]
```diff
@@ -266,132 +286,104 @@ namespace MueLu {
...
+
     const LO  numRows = graph.GetNodeNumVertices();
     const int myRank  = graph.GetComm()->getRank();
 
-    ArrayRCP<LO> vertex2AggId = aggregates.GetVertex2AggId()->getDataNonConst(0);
-    ArrayRCP<LO> procWinner   = aggregates.GetProcWinner()  ->getDataNonConst(0);
+    auto vertex2AggIdView = aggregates.GetVertex2AggId()->template getLocalView<memory_space>();
+    auto procWinnerView = aggregates.GetProcWinner()    ->template getLocalView<memory_space>();
 
+    LO numNonAggregatedNodesSerial = numNonAggregatedNodes;
```
Comment (brian-kelley) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase1Algorithm_kokkos_def.hpp]: I'm getting an declared but unused warning for numNonAggregatedNodesSerial (CUDA build on white)

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase3Algorithm_kokkos_def.hpp [hunk 2]
```diff
@@ -65,88 +65,123 @@ namespace MueLu {
...
+                               if (j < as<int>(neighOfINode.length)) {
+                                 // Assign to an adjacent aggregate
+                                 vertex2AggId(i, 0) = vertex2AggId(neighOfINode(j), 0);
+                               } else if (error_on_isolated) {
+                                 // Error on this isolated node, as the user has requested
+                                 std::ostringstream oss;
+                                 oss<<"MueLu::AggregationPhase3Algorithm::BuildAggregates: MueLu has detected a non-Dirichlet node that has no on-rank neighbors and is terminating (by user request). "<<std::endl;
+                                 oss<<"If this error is being generated at level 0, this is due to an initial partitioning problem in your matrix."<<std::endl;
+                                 oss<<"If this error is being generated at any other level, try turning on repartitioning, which may fix this problem."<<std::endl;
+                                 throw Exceptions::RuntimeError(oss.str());
```
Comment (brian-kelley) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase3Algorithm_kokkos_def.hpp]: Throwing in device code is giving me a compile error with CUDA. Would it be possible to enable the error checking and output in host space only? I'm not sure of the most elegant way to do that.

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp [hunk 3]
```diff
@@ -90,57 +114,85 @@ namespace MueLu {
...
+                                  auto neighOfINode = graph.getNeighborVertices(vertexIdx);
+
+                                  for (int j = 0; j < as<int>(neighOfINode.length); j++) {
+                                    LO neigh = neighOfINode(j);
+
+                                    // We don't check (neigh != i), as it is covered by checking
+                                    // (aggStat[neigh] == AGGREGATED)
+                                    if ( graph.isLocalNeighborVertex(neigh)
+                                         && (aggStatView(neigh) == AGGREGATED) )
+                                      aggWeightView(j) = aggWeightView(j)
```
Comment (jhux2) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp]: Index j is wrong.  Instead, we want to the weight to be computed per aggregate, not per individual nbr.

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp [hunk 4]
```diff
@@ -90,57 +114,85 @@ namespace MueLu {
...
+                                        + connectWeightView(neigh);
+                                  }
+
+                                  int bestScore   = -100000;
+                                  int bestAggId   = -1;
+                                  int bestConnect = -1;
+
+                                  for (int j = 0; j < as<int>(neighOfINode.length); j++) {
```
Comment (jhux2) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp]: Can we rename INode to something more meaningful?

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp [hunk 5]
```diff
@@ -90,57 +114,85 @@ namespace MueLu {
...
+                                    LO neigh = neighOfINode(j);
+
+                                    if ( graph.isLocalNeighborVertex(neigh)
+                                         && (aggStatView(neigh) == AGGREGATED) ) {
+                                      int aggId = vertex2AggIdView(neigh, 0);
+                                      int score = aggWeightView(j) - aggPenaltiesView(aggId);
```
Comment (jhux2) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp]: j should an aggregate index

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase1Algorithm_kokkos_def.hpp (hunk 1 above)
Comment (lucbv) [reply packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase1Algorithm_kokkos_def.hpp]: @brian-kelley: OK, I took care of that, you should not have any problem compiling any more.
I have also found a good solution for the memory access problem. I am now working on phase 2b aggregation where I am allocating too much L0 memory apparently...

**Comment by lucbv** [comment]
@brian-kelley this last commit has fixed the compilation and memory issues. Everything should now run fine. I still need to reply to some of the question by @jhux2, most importantly there is still a bug in Phase2b!

**Comment by lucbv** [comment]
@jhux2 I pushed a fix in phase 2b for the bug that we detected, it should now accumulate values per aggregate correctly when trying to add a node to an aggregate.
The main change is a new loop over the fine node neighbors that detect the aggregate id of the neighbors and assign that aggregate a new "local id". This local id is then used for values accumulation such as aggregate weight.

I am also noticing that I should probably have the view called `neighOfINode` be created in scratch memory as well as pushing the creation of `vertex2AggLIDView` and `aggWeightView` inside the `if (aggStatView(vertexIdx) == READY)` statement to avoid creating them unnecessarily if the node is not `READY` ...

**Comment by aprokop** [comment]
I looked briefly through the code. Phase 1 and 2a look fine. Phase 2b looks a bit too complicated, and I'm still uncertain about some details. Phase 3 seems easy. I assume that the code will go through further cleanup?

Is it worth running phase 2b on the device?

**Comment by lucbv** [comment]
Thanks for looking at, I am planning on pushing two more commits soon, first I will add separate treatments for the aggregates if `maxAggSize` is specified or not, since in the non-specified case a few constraint can be lifted on the algorithm which should speed things up, second I will explore the opportunity of adding more fine grained parallelism using team policies, part of this is already implemented locally but I want to modify my commits to allow for an easy revert of the team policies in case they do not speed up the algorithm.

Regarding `phase 2b`, I think that it is even more important now since in quite a few test problems I have observed that a lot of work is done in phase 1, none in phase 2a and then a good amount in phase 2b and usually very little to no work happens in phase 3.

One of the main complication in phase 2b compared to the serial implementation is the memory management. If you want to talk about this we can meet on the phone.

**Comment by aprokop** [comment]
Just ran my comparison driver for old and new aggregation using a single thread with OpenMP. On a Laplace3D problem size 150^3:
```
timer name                                                                                                 old1.log.yaml new1.log.yaml      ratio
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Build (total, level=0)                                    0.238      0.933       3.92
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Build (level=0)                                           0.238      0.933       3.92
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 1 (main)" (sub, total, level=0)               0.092      0.089       0.96
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 2b (expansion)" (sub, total, level=0)         0.082      0.075       0.92
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 2a (secondary)" (sub, total, level=0)         0.029      0.214       7.42
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 3 (cleanup)" (sub, total, level=0)            0.003      0.010       3.78
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase - (Dirichlet)" (sub, total, level=0)          0.002      0.005       2.65
```
So,
a) 2a seems to be a lot slower
b) some time is not accounted for which leads to dramatic increase in overall aggregation (x4).

**Comment by aprokop** [comment]
Comparing single thread with multiple shows little scaling:
```
timer name                                                                                                 new4.log.yaml new1.log.yaml      ratio
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Build (total, level=0)                                    0.584      0.753       1.29
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Build (level=0)                                           0.584      0.753       1.29
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Coloring (sub, total, level=0)                            0.290      0.309       1.07
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 2a (secondary)" (sub, total, level=0)         0.088      0.227       2.58
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 1 (main)" (sub, total, level=0)               0.070      0.070       1.00
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 2b (expansion)" (sub, total, level=0)         0.049      0.085       1.74
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase - (Dirichlet)" (sub, total, level=0)          0.012      0.003       0.25
MueLu: Laplace3D: UncoupledAggregationFactory_kokkos: Algo "Phase 3 (cleanup)" (sub, total, level=0)            0.009      0.009       1.00
```
It is possible that I may be doing something wrong. @lucbv Do do you have any performance results with the new code?

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp [hunk 6]
```diff
@@ -90,57 +114,85 @@ namespace MueLu {
     // Assuming that the aggregate size is 3 in each direction running the algorithm only twice
     // should be sufficient.
```
[the remark this answers is not in the document]
Comment (lucbv) [reply packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp]: Added a little piece of code that modifies the number of iterations based on a simple check of the value stored in the parameter list.

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2aAlgorithm_kokkos_def.hpp [hunk 7]
```diff
@@ -63,78 +63,108 @@
...
+    typename Kokkos::View<LO, memory_space>::HostMirror h_numLocalAggregates =
+      Kokkos::create_mirror_view(numLocalAggregates);
+    h_numLocalAggregates() = aggregates.GetNumAggregates();
+    Kokkos::deep_copy(numLocalAggregates, h_numLocalAggregates);
+
+    // Now we create new aggregates using root nodes in all colors other than the first color,
+    // as the first color was already exhausted in Phase 1.
+    Kokkos::TeamPolicy<execution_space> outerPolicy(numRows, Kokkos::AUTO);
+    typedef typename Kokkos::TeamPolicy<execution_space>::member_type  member_type;
+    for(int color = 1; color < numColors + 1; ++color) {
```
[the remark this answers is not in the document]
Comment (lucbv) [reply packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2aAlgorithm_kokkos_def.hpp]: @brian-kelley  yes this should be a two there, I had in mind that colors are starting at 0 but that's actually 1, I'll make that modification now.
I will work a bit on phase 3 today since it is way to slow on GPU which is annoying...

**Comment by lucbv** [comment]
@srajama1 there is no doc for the design changes at this point. I am however running some tests on white and bowman at various stages of development to make sure that we can revert some commits if we feel things need to change.

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2aAlgorithm_kokkos_def.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (lucbv) [reply packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2aAlgorithm_kokkos_def.hpp]: @brian-kelley  actually I remembered that you are working on phase 3 already so I will leave as is for now until we decide to merge things together.

**Comment by srajama1** [comment]
@lucbv : Thanks, I was just checking to see, no worries.
